src/fs/file.rs

```diff
@@ -514,6 +514,58 @@ impl<'dir> File<'dir> {
         }
     }
 
+    /// Recursive folder size
+    #[cfg(unix)]
+    pub fn recursive_size(&self) -> f::Size {
+        use crate::fs::RECURSIVE_SIZE_HASHMAP;
+        if self.is_directory() {
+            let Ok(dir) = Dir::read_dir(self.path.clone()) else {
+                return f::Size::None;
+            };
+            let files = dir.files(super::DotFilter::Dotfiles, None, false, false);
+
+            let mut recursive_size: u64 = 0;
+            for fileresult in files {
+                let Ok(file) = fileresult else { continue };
+                if file.is_file() {
+                    recursive_size += file.metadata.size();
+                } else {
+                    recursive_size += match file.recursive_size() {
+                        f::Size::Some(s) => s,
+                        _ => file.metadata.size(),
+                    };
+                }
+            }
+            RECURSIVE_SIZE_HASHMAP
+                .lock()
+                .unwrap()
+                .insert(self.metadata.ino(), recursive_size);
```

Comment:
What happens if we cross filesystem boundaries and have the same inode number?  Do we want to handle that case?

Reply:
Also we should only be using the cache if it is a recursive directory listing and it is a directory and --total-size is used.  If we are storing file sizes it is just wasting a lot of space.